### PR TITLE
Add view_slice ttgir instruction

### PR DESCRIPTION
**Pull request by oplavsic**
This instruction represents view of the slice of the tensor in registers. Syntax of the operation is the same as for extract_slice op. However, unlike 'extract_slice' which slices in shared memory, 'view_slice' specifically slices within registers.
Slice of the tensor is required to have the same layout as the original tensor. In a way, semantics of the 'view_slice' operation is a combination of the 'extract_slice' and 'view' operations semantics.

This PR adds support for slicing tensors in blocked layout.
TODO: Support slicing in dot layout.

### CODE REVIEW

File: lib/Conversion/TritonGPUToLLVM/TritonGPUToLLVM.cpp [hunk 1]
```diff
@@ -685,6 +685,128 @@ struct ExtractSliceOpConversion
   }
 };
 
+/***
+   # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # #
```
Comment (oplavsic) [inline lib/Conversion/TritonGPUToLLVM/TritonGPUToLLVM.cpp:689]: Diff shows this graphical representation in a bad format. You can open the file to see it properly.
I can remove this representation if you think it's too much, but it should at least make review easier.

File: include/triton/Dialect/TritonGPU/IR/TritonGPUOps.td [hunk 2]
```diff
@@ -227,6 +227,67 @@ def TTG_InsertSliceOp : TTG_Op<"insert_slice",
 }
 
 
+def TTG_ViewSliceOp : TTG_Op<"view_slice",
```
Comment (oplavsic) [inline include/triton/Dialect/TritonGPU/IR/TritonGPUOps.td:230]: I retained syntax of the extract_slice operation, so this td description is (almost) the same as for extract_slice.

File: include/triton/Dialect/TritonGPU/IR/TritonGPUOps.td [hunk 3]
```diff
@@ -227,6 +227,67 @@ def TTG_InsertSliceOp : TTG_Op<"insert_slice",
...
+                                [AttrSizedOperandSegments,
+                                //  ResultsAreSharedEncoding,
```
Comment (zhanglx13) [inline include/triton/Dialect/TritonGPU/IR/TritonGPUOps.td:232]: I'm thinking if we want to add a trait to check that the slice and the parent tensor can have the same layout.

File: include/triton/Dialect/TritonGPU/IR/TritonGPUOps.td [hunk 4]
```diff
@@ -227,6 +227,67 @@ def TTG_InsertSliceOp : TTG_Op<"insert_slice",
...
+    /// Returns the type of the base tensor operand.
+    RankedTensorType getSourceType() {
+      return getSource().getType().cast<RankedTensorType>();
+    }
+
+    std::array<unsigned, 3> getArrayAttrMaxRanks() {
+      unsigned rank = getSourceType().getRank();
+      return {rank, rank, rank};
+    }
+  }];
```
Comment (zhanglx13) [inline include/triton/Dialect/TritonGPU/IR/TritonGPUOps.td:279]: These are for `extract_slice` op. Just a note that when you actually use the `view_slice` op, you may want to have something different.

File: python/test/unit/language/test_core_amd.py [hunk 5]
```diff
@@ -2917,6 +2917,71 @@ def test_convert2d(dtype, shape, src_layout, dst_layout, device='cuda'):
...
+    %0 = tt.make_range {{end = {M} : i32, start = 0 : i32}} : tensor<{M}xi32, #triton_gpu.slice<{{dim = 1, parent = #src}}>>
+    %42 = tt.make_range {{end = {M_tile_size} : i32, start = 0 : i32}} : tensor<{M_tile_size}xi32, #triton_gpu.slice<{{dim = 1, parent = #src}}>>
+    %1 = tt.make_range {{end = {M} : i32, start = 0 : i32}} : tensor<{M}xi32, #triton_gpu.slice<{{dim = 0, parent = #src}}>>
+    %2 = tt.splat %arg0 : (!tt.ptr<f16>) -> tensor<{M}x{N}x!tt.ptr<f16>, #src>
+    %4 = tt.expand_dims %0 {{axis = 1 : i32}} : (tensor<{M}xi32, #triton_gpu.slice<{{dim = 1, parent = #src}}>>) -> tensor<{M}x1xi32, #src>
+    %43 = tt.expand_dims %42 {{axis = 1 : i32}} : (tensor<{M_tile_size}xi32, #triton_gpu.slice<{{dim = 1, parent = #src}}>>) -> tensor<{M_tile_size}x1xi32, #src>
+    %5 = arith.muli %4, %cst : tensor<{M}x1xi32, #src>
+    %44 = arith.muli %43, %cst_n : tensor<{M_tile_size}x1xi32, #src>
+    %6 = tt.expand_dims %1 {{axis = 0 : i32}} : (tensor<{M}xi32, #triton_gpu.slice<{{dim = 0, parent = #src}}>>) -> tensor<1x{M}xi32, #src>
+    %7 = tt.broadcast %6 : (tensor<1x{M}xi32, #src>) -> tensor<{M}x{M}xi32, #src>
```
Comment (zhanglx13) [inline python/test/unit/language/test_core_amd.py:2950]: Several places: {M}x{M} should be {M}x{N}

File: python/test/unit/language/test_core_amd.py [hunk 6]
```diff
@@ -2917,6 +2917,71 @@ def test_convert2d(dtype, shape, src_layout, dst_layout, device='cuda'):
...
 
 
+layouts = [
+    BlockedLayout([2, 2], [16, 4], [2, 2], [1, 0], [1, 1], [1, 1], [0, 1]),
+    BlockedLayout([2, 2], [16, 4], [2, 2], [0, 1], [1, 1], [1, 1], [0, 1]),
+    BlockedLayout([1, 8], [16, 4], [4, 1], [1, 0], [1, 1], [1, 1], [0, 1]),
+    BlockedLayout([1, 8], [16, 4], [4, 1], [0, 1], [1, 1], [1, 1], [0, 1]),
+]
+
+@pytest.mark.parametrize("M, N, M_tile_size, N_tile_size, M_tile_offset, N_tile_offset", [[256, 256, 128, 64, 64, 128], [128, 128, 128, 32, 0, 0], [128, 128, 128, 32, 0, 64]])
```
Comment (zhanglx13) [inline python/test/unit/language/test_core_amd.py:2927]: I tried to add 
`[256, 128, 256, 32, 0, 0]`
And fixed all {M}x{M} issues in the IR, but there are mismatches for this config. Can you take a look?

**Review by zhanglx13 — CHANGES_REQUESTED** [review]
The implementation seems good to me. However, there are mismatches for an untested config. See comment below.

**Review by zhanglx13 — APPROVED** [review]

**Pull request merged**